### PR TITLE
Algorithm: N-1 pareto optimal profiles

### PR DESCRIPTION
**Pull request by Tristramg**

### CODE REVIEW

File: src/algo.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+fn best_connection(profiles: &Vec<Profile>, c: &Connection) -> Option<Profile> {
+    let transfer_duration = 5;
+    profiles
+        .iter()
+        .take_while(|p| p.dep_time > c.arr_time + transfer_duration)
+        .last()
+        .map(|p| {
+            Profile {
```
Comment (Tristramg) [inline src/algo.rs]: How should I make a copy of the Profile p?

File: src/algo.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            in_connection: None,
+            dep_time: u16::max_value(),
+            arr_time: 0,
+        }
+    }
+
+    fn domination(&self, other_option: Option<&Self>) -> Domination {
+        use std::cmp::Ordering::*;
+        if self.arr_time == u16::max_value() {
+            Domination::Dominated
```
Comment (TeXitoi) [inline src/algo.rs]: You mean dep_time?

File: src/algo.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        } else if let Some(other) = other_option {
+            match (
+                self.dep_time.cmp(&other.dep_time),
+                self.arr_time.cmp(&other.arr_time),
+            ) {
+                (Equal, Equal) => Domination::Equal,
+                (Less, Less) | (Greater, Greater) => Domination::Equivalent,
+                (Greater, _) | (Equal, Less) => Domination::Dominates,
+                (Less, _) | (Equal, Greater) => Domination::Dominated,
```
Comment (TeXitoi) [inline src/algo.rs]: These Equal are a bit disturbing as they add nothing, who about jut put _ ?

File: src/algo.rs (hunk 2 above)
Comment (TeXitoi) [reply src/algo.rs]: I don't get this special case.

File: src/algo.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            }
+        } else {
+            Domination::Dominates
+        }
+    }
+}
+
+fn arrival_time_with_stop_change(profiles: &Vec<Profile>, c: &Connection) -> u16 {
```
Comment (TeXitoi) [inline src/algo.rs]: We never use &Vec<_> we use &[_]

File: src/algo.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    let transfer_duration = 5;
+    profiles
+        .iter()
+        .take_while(|p| p.dep_time > c.arr_time + transfer_duration)
+        .last()
+        .map_or(u16::max_value(), |p| {
```
Comment (TeXitoi) [inline src/algo.rs]: Special case because of that?

File: src/algo.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                p.arr_time
+            } else {
+                // If this is the very last connection to target, it gives us the arrival time
+                c.arr_time
+            }
+        })
+}
+
+// It returns all the possible routes, from all possible nodes to the given destination
+pub fn compute(timetable: Timetable, destination: usize) -> Vec<Vec<Profile>> {
```
Comment (TeXitoi) [inline src/algo.rs]: &Timetable

File: src/algo.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    let mut trips: Vec<_> = timetable.trips.iter().map(|_| u16::max_value()).collect();
```
Comment (TeXitoi) [inline src/algo.rs]: Badly named, looks like a date time corresponding to a trip.

File: src/algo.rs (hunk 5 above)
Comment (Tristramg) [reply src/algo.rs]: yes… I’m not very happy with this solution. I tried to fiddle around keeping the option, but it got way to clumsy

File: src/algo.rs (hunk 1 above)
Comment (TeXitoi) [reply src/algo.rs]: Derive clone (or copy) and you can clone()

File: src/algo.rs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+impl Profile {
+    fn default() -> Self {
+        Self {
+            in_connection: None,
+            dep_time: u16::max_value(),
+            arr_time: 0,
+        }
+    }
```
Comment (TeXitoi) [inline src/algo.rs]: Better to impl Default

**Comment by Tristramg** [comment]
I rewrote quite a bit. I think it is clearer now.

No longer max_values to handle.

The _incorporate_ functions seems a bit clumsy, but I couldn’t find how to make it nicer :(

File: src/algo.rs (hunk 8 above)
Comment (Tristramg) [reply src/algo.rs]: so many traits, so little time ;) :+1:

File: src/algo.rs [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+            out_connection: None,
+            dep_time: u16::max_value(),
+            arr_time: 0,
+        }
+    }
+}
+
+impl Profile {
+    fn dominates(&self, other: &Self) -> bool {
+        self.arr_time <= other.arr_time && self.arr_time >= other.dep_time
```
Comment (antoine-de) [inline src/algo.rs]: isn't it `&& self.dep_time >= other.dep_time` ?

File: src/algo.rs [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    fn is_non_dominated(&self, other_opt: Option<&Self>) -> bool {
+        // By construction, we know self.dep_time >= other.dep_time
+        match other_opt {
+            Some(other) => self.arr_time <= other.arr_time,
```
Comment (antoine-de) [inline src/algo.rs]: I don't understand, if it's `self.arr_time <= other.arr_time` and `self.dep_time >= other.dep_time` isn't it the same as `dominates` (maybe I did not correctly understood `dominates` though) ?

File: src/algo.rs [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+        // Case 2: Staying seated in the trip, we will reach the target at `t2`
+        let t2 = arr_time_with_trip[c.trip];
+
+        // Case 3: Transfering in the same stop, we look up the earliest compatible arrival
+        let t3 = arrival_time_with_stop_change(&profiles[c.arr_stop], c);
+
+        if let Some(t) = [t2, t3].iter().filter_map(|t| *t).min_by(|a, b| a.cmp(b)) {
+            let candidate = Profile {
+                out_connection: Some(conn_index),
```
Comment (antoine-de) [inline src/algo.rs]: what do you use out_connection for ? it seems to be only used as a boolean (I don't like the untyped index)

File: src/algo.rs [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+        assert_eq!(1, profiles.len());
+
+        profiles.incorporate(Profile {
+            dep_time: 10,
+            arr_time: 20,
+            out_connection: None,
+        });
+        assert_eq!(2, profiles.len());
+
+        // Dominated profile, should
```
Comment (antoine-de) [inline src/algo.rs]: truncated line

File: src/algo.rs (hunk 9 above)
Comment (Tristramg) [reply src/algo.rs]: :no_mouth:

File: src/algo.rs (hunk 10 above)
Comment (Tristramg) [reply src/algo.rs]: It’s a typo in the comment. other.dep_time >= self.dep_time
Thanks

File: src/algo.rs (hunk 11 above)
Comment (Tristramg) [reply src/algo.rs]: It will be used later to reconstruct the path, for now it’s indeed no really needed yet

File: src/algo.rs (hunk 12 above)
Comment (Tristramg) [reply src/algo.rs]: a problem with my keyb

Comment (antoine-de) [reply src/algo.rs]: :rofl:

**Review by antoine-de — APPROVED** [review]

**Pull request merged**